bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -3251,6 +3268,41 @@ void ReplicaImp::onMessage<SimpleAckMsg>(SimpleAckMsg *msg) {
   delete msg;
 }
 
+template <>
+void ReplicaImp::onMessage<ReplicaRestartReadyMsg>(ReplicaRestartReadyMsg *msg) {
+  LOG_INFO(GL,
+           "Recieved  ReplicaRestartReadyMsg from sender_id " << std::to_string(msg->idOfGeneratedReplica())
+                                                              << " with seq_num" << std::to_string(msg->seqNum()));
+  restart_ready_list_.emplace_back(std::make_unique<ReplicaRestartReadyMsg>(msg));
+  bool bft = bftEngine::ControlStateManager::instance().getRestartBftFlag();
+  uint32_t targetNumOfMsgs = (bft ? (config_.getnumReplicas() - config_.getfVal()) : config_.getnumReplicas());
+  if (restart_ready_list_.size() == targetNumOfMsgs) {
+    LOG_INFO(GL, "Target number = " << targetNumOfMsgs << " of restart ready msgs are recieved. Send resatrt proof");
+    sendReplicasRestartReadyProof();
```

Comment:
Once the replica creates and sends the proof it can restart itself (I'm pretty sure it doesn't send the message to itself)